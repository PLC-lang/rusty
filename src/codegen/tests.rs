// Copyright (c) 2020 Ghaith Hachem and Mathias Rieder
mod code_gen_tests;
mod codegen_error_messages_tests;
mod compare_instructions_tests;
mod constants_tests;
mod debug_tests;
mod directaccess_test;
mod expression_tests;
mod function_tests;
mod generics_test;
mod initialization_test;
mod parameters_tests;
mod statement_codegen_test;
mod string_tests;
#[cfg(feature = "verify")]
mod switch_case_tests;
mod typesystem_test;
