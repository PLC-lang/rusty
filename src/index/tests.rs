// Copyright (c) 2020 Ghaith Hachem and Mathias Rieder
mod generic_tests;
mod index_tests;
mod instance_resolver_tests;
