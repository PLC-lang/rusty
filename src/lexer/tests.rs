/// Copyright (c) 2020 Ghaith Hachem and Mathias Rieder

use pretty_assertions::{assert_eq, assert_ne};
#[test]
fn pou_tokens() {
    let mut lexer = super::lex("PROGRAM END_PROGRAM FUNCTION END_FUNCTION FUNCTION_BLOCK END_FUNCTION_BLOCK");
    assert_eq!(lexer.token, super::Token::KeywordProgram);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordEndProgram);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordFunction);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordEndFunction);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordFunctionBlock);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordEndFunctionBlock);
    lexer.advance();
}

#[test]
fn var_tokens() {
    let mut lexer = super::lex("VAR VAR_INPUT VAR_GLOBAL END_VAR");
    assert_eq!(lexer.token, super::Token::KeywordVar);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordVarInput);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordVarGlobal);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordEndVar);
}

#[test]
fn hello_is_an_identifier() {
    let mut lexer = super::lex("hello a12 _a12");
    assert_eq!(lexer.token, super::Token::Identifier, "{}", lexer.slice());
    lexer.advance();
    assert_eq!(lexer.token, super::Token::Identifier, "{}", lexer.slice());
    lexer.advance();
    assert_eq!(lexer.token, super::Token::Identifier, "{}", lexer.slice());
    lexer.advance();
}

#[test]
fn an_identifier_cannot_start_with_a_number() {
    let lexer = super::lex("2g12");
    assert_ne!(lexer.token, super::Token::Identifier);
}

#[test]
fn punctuations() {
    let lexer = super::lex(":");
    assert_eq!(lexer.token, super::Token::KeywordColon, "{}", lexer.slice());
    let lexer = super::lex(";");
    assert_eq!(
        lexer.token,
        super::Token::KeywordSemicolon,
        "{}",
        lexer.slice()
    );
}

#[test]
fn parens() {
    let mut lexer = super::lex("( )");
    assert_eq!(lexer.token, super::Token::KeywordParensOpen);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordParensClose);
}

#[test]
fn a_assignment_is_keword_assignment() {
    let lexer = super::lex(":=");
    assert_eq!(lexer.token, super::Token::KeywordAssignment);
}

#[test]
fn comma(){
    let lexer = super::lex(",");
    assert_eq!(lexer.token, super::Token::KeywordComma);
}

#[test]
fn operator_test() {
    let mut lexer = super::lex("+ - * / MOD = <> < > <= >=");
    assert_eq!(lexer.token, super::Token::OperatorPlus);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::OperatorMinus);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::OperatorMultiplication);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::OperatorDivision);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::OperatorModulo);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::OperatorEqual);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::OperatorNotEqual);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::OperatorLess);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::OperatorGreater);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::OperatorLessOrEqual);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::OperatorGreaterOrEqual);
}

#[test]
fn boolean_expression_test() {
    let mut lexer = super::lex("AND XOR OR NOT");
    assert_eq!(lexer.token, super::Token::OperatorAnd);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::OperatorXor);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::OperatorOr);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::OperatorNot);
}

#[test]
fn int_literals_test() {
    let mut lexer = super::lex("1 2 3 0123 321");

    for x in 0..5 {
        print!("{}", x);
        assert_eq!(lexer.token, super::Token::LiteralInteger);
        lexer.advance();
    }
}

#[test]
fn real_literals_test() {
    let mut lexer = super::lex("1.234 0.9E10");

        assert_eq!(lexer.token, super::Token::LiteralInteger);
        lexer.advance();
        assert_eq!(lexer.token, super::Token::KeywordDot);
        lexer.advance();
        assert_eq!(lexer.token, super::Token::LiteralInteger);
        lexer.advance();
        assert_eq!(lexer.token, super::Token::LiteralInteger);
        lexer.advance();
        assert_eq!(lexer.token, super::Token::KeywordDot);
        lexer.advance();
        assert_eq!(lexer.token, super::Token::LiteralInteger);
        lexer.advance();
        assert_eq!(lexer.token, super::Token::LiteralExponent);
        lexer.advance();
}

#[test]
fn a_full_program_generates_correct_token_sequence() {
    let mut lexer = super::lex(
        r"
        PROGRAM hello
        VAR
          a : INT;
        END_VAR
        END_PROGRAM
        ",
    );

    assert_eq!(lexer.token, super::Token::KeywordProgram);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::Identifier);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordVar);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::Identifier);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordColon);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::Identifier);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordSemicolon);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordEndVar);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordEndProgram);
}

#[test]
fn boolean_literals() {
    let mut lexer = super::lex(r" TRUE FALSE");
    assert_eq!(lexer.token, super::Token::LiteralTrue);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::LiteralFalse);
}

#[test]
fn if_expression() {
    let mut lexer = super::lex(
        r"
        IF THEN ELSIF ELSE END_IF
        ",
    );

    assert_eq!(lexer.token, super::Token::KeywordIf);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordThen);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordElseIf);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordElse);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordEndIf);
}

#[test]
fn for_statement() {
    let mut lexer = super::lex(
        r"
        FOR TO BY DO END_FOR
        ",
    );

    assert_eq!(lexer.token, super::Token::KeywordFor);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordTo);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordBy);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordDo);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordEndFor);
}

#[test]
fn while_statement() {
    let mut lexer = super::lex(
        r"
        WHILE DO END_WHILE
        ",
    );

    assert_eq!(lexer.token, super::Token::KeywordWhile);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordDo);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordEndWhile);
}

#[test]
fn repeat_statement() {
    let mut lexer = super::lex(
        r"
        REPEAT UNTIL END_REPEAT
        ",
    );

    assert_eq!(lexer.token, super::Token::KeywordRepeat);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordUntil);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordEndRepeat);
}

#[test]
fn case_statement() {
    let mut lexer = super::lex(
        r"
        CASE OF ELSE END_CASE
        ",
    );

    assert_eq!(lexer.token, super::Token::KeywordCase);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordOf);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordElse);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordEndCase);
}

#[test]
fn dot_statements() {
    let mut lexer = super::lex(
        r".. .",
    );

    assert_eq!(lexer.token, super::Token::KeywordDotDot);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordDot);
    lexer.advance();
}

#[test]
fn range_statements() {
    let mut lexer = super::lex(
        r"123..ABC",
    );

    println!("{:?}", lexer.token);
    lexer.advance();
    println!("{:?}", lexer.token);
    lexer.advance();
    println!("{:?}", lexer.token);
    lexer.advance();
}

#[test]
fn struct_enum_datatype() {
    let mut lexer = super::lex(
        r"TYPE STRUCT END_STRUCT END_TYPE",
    );

    assert_eq!(lexer.token, super::Token::KeywordType);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordStruct);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordEndStruct);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordEndType);
}

#[test]
fn array_parsing() {
    let mut lexer = super::lex(
        r"ARRAY OF x[5]",
    );

    assert_eq!(lexer.token, super::Token::KeywordArray);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordOf);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::Identifier);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordSquareParensOpen);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::LiteralInteger);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordSquareParensClose);
    lexer.advance();
    
}

#[test]
fn string_parsing() {
    let mut lexer = super::lex(
        r"'AB C' 'AB$$' 'AB$''"
    );

    assert_eq!(lexer.token, super::Token::LiteralString);
    assert_eq!("'AB C'",lexer.slice());
    lexer.advance();
    assert_eq!(lexer.token, super::Token::LiteralString);
    assert_eq!("'AB$$'",lexer.slice());
    lexer.advance();
    assert_eq!(lexer.token, super::Token::LiteralString);
    assert_eq!("'AB$''",lexer.slice());
    lexer.advance();
}
    
