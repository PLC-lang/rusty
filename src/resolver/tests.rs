use crate::{
    ast::{self, CompilationUnit},
    index::{self, Index},
    lexer::lex,
};

use super::{AnnotationMap, TypeAnnotator};

#[cfg(test)]
mod const_resolver_tests;
#[cfg(test)]
mod resolve_control_statments;
#[cfg(test)]
mod resolve_expressions_tests;
#[cfg(test)]
mod resolve_literals_tests;

fn parse(src: &str) -> (CompilationUnit, Index) {
    let (mut unit, _) = crate::parser::parse(lex(src));

    ast::pre_process(&mut unit);
    let index = index::visitor::visit(&unit);
    (unit, index)
}

fn annotate(parse_result: &CompilationUnit, index: &Index) -> AnnotationMap {
    TypeAnnotator::visit_unit(index, parse_result)
}
