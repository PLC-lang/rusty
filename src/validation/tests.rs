// Copyright (c) 2020 Ghaith Hachem and Mathias Rieder
mod array_access_validation_test;
mod bitaccess_validation_test;
mod literals_validation_tests;
mod reference_resolve_tests;
mod variable_validation_tests;

mod statement_validation_tests;

mod generic_validation_tests;
