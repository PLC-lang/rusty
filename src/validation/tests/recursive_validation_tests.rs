mod edgecases {
    use crate::test_utils::tests::parse_and_validate;

    #[test]
    fn pointers_should_not_be_considered_as_cycle() {
        let diagnostics = parse_and_validate(
            "
            TYPE A : STRUCT
                b : B;
            END_STRUCT END_TYPE

            TYPE B : STRUCT
                a : REF_TO A;
            END_STRUCT END_TYPE
            ",
        );

        assert_eq!(diagnostics.len(), 0);
    }

    // From https://github.com/PLC-lang/rusty/pull/748:
    // Running cargo r -- ../standardfunctions/iec61131-st/*.st previously returned a weird TIME -> TIME
    // recursion which shouldn't happen. Instead of spending time debugging that one edge-case we now
    // explicitly filter for nodes within the dfs method. As a nice-to-have this is probably also more performant.
    //
    // This test covers the above edge-case
    #[test]
    fn external_function_should_not_trigger() {
        let diagnostics = parse_and_validate(
            "
            {external}
            FUNCTION TIME : TIME
            END_FUNCTION

            TYPE niceTimes : STRUCT
                x : TIME;
                y : DATE_AND_TIME;
            END_STRUCT
            END_TYPE
            ",
        );

        assert_eq!(diagnostics.len(), 0);
    }

    #[test]
    fn struct_and_function_with_same_name() {
        let diagnostics = parse_and_validate(
            "
            FUNCTION TM : TM
            END_FUNCTION

            TYPE TM : STRUCT
                hours, minutes, seconds : DINT;
            END_STRUCT
            END_TYPE
            ",
        );

        assert_eq!(diagnostics.len(), 0);
    }

    #[test]
    fn struct_and_function_with_same_name_2() {
        let diagnostics = parse_and_validate(
            "
            FUNCTION TM : DINT
                VAR_INPUT
                    x : TM;
                END_VAR
            END_FUNCTION

            TYPE TM : STRUCT
                hours, minutes, seconds : DINT;
            END_STRUCT
            END_TYPE
            ",
        );

        assert_eq!(diagnostics.len(), 0);
    }

    #[test]
    fn struct_and_function_with_same_name_3() {
        let diagnostics = parse_and_validate(
            "
            FUNCTION TM : DINT
                VAR_INPUT
                    x : TM;
                END_VAR
            END_FUNCTION

            TYPE TM : STRUCT
                TM : DINT;
            END_STRUCT
            END_TYPE
            ",
        );

        assert_eq!(diagnostics.len(), 0);
    }
}

mod structs {
    use insta::assert_snapshot;

    use crate::{test_utils::tests::parse_and_validate, validation::tests::make_readable};

    #[test]
    fn one_cycle_abca() {
        let diagnostics = parse_and_validate(
            "
            TYPE A : STRUCT
                b : B;
            END_STRUCT END_TYPE

            TYPE B : STRUCT
                c : C;
            END_STRUCT END_TYPE

            TYPE C : STRUCT
                a : A;
                e : e;
            END_STRUCT END_TYPE
            
            TYPE E : STRUCT
                a_int: INT;
            END_STRUCT END_TYPE
            ",
        );

        assert_snapshot!(make_readable(&diagnostics));
    }

    #[test]
    fn one_cycle_self_a() {
        let diagnostics = parse_and_validate(
            "
            TYPE A : STRUCT
                a : A;
            END_STRUCT END_TYPE
            ",
        );

        assert_snapshot!(make_readable(&diagnostics));
    }

    #[test]
    fn one_cycle_multiple_self_a() {
        let diagnostics = parse_and_validate(
            "
            TYPE A : STRUCT
                a1 : A;
                a2 : A;
                a3 : A;
            END_STRUCT END_TYPE
            ",
        );

        assert_snapshot!(make_readable(&diagnostics));
    }

    #[test]
    fn one_cycle_aba() {
        let diagnostics = parse_and_validate(
            "
            TYPE A : STRUCT
                b : B;
            END_STRUCT END_TYPE

            TYPE B : STRUCT
                a : A;
            END_STRUCT END_TYPE
            ",
        );

        assert_snapshot!(make_readable(&diagnostics));
    }

    #[test]
    fn one_cycle_bcb() {
        let diagnostics = parse_and_validate(
            "
            TYPE A : STRUCT
                b : B;
            END_STRUCT END_TYPE
            
            TYPE B : STRUCT
                c : C;
            END_STRUCT END_TYPE
            
            TYPE C : STRUCT
                b : B;
            END_STRUCT END_TYPE
            ",
        );

        assert_snapshot!(make_readable(&diagnostics));
    }

    #[test]
    fn one_cycle_with_multiple_identical_members_aba() {
        let diagnostics = parse_and_validate(
            "
            TYPE A : STRUCT 
                b1 : B;
                b2 : B;
                b3 : B;
            END_STRUCT END_TYPE

            TYPE B : STRUCT 
                a : A;
            END_STRUCT END_TYPE
            ",
        );

        assert_snapshot!(make_readable(&diagnostics));
    }

    #[test]
    fn two_cycles_aa_and_aba() {
        let diagnostics = parse_and_validate(
            "
            TYPE A : STRUCT
                a : A;
                b : B;
            END_STRUCT END_TYPE
            
            TYPE B : STRUCT
                a : A;
            END_STRUCT END_TYPE
            ",
        );

        assert_snapshot!(make_readable(&diagnostics));
    }

    #[test]
    fn two_cycles_branch_cc_and_cec() {
        let diagnostics = parse_and_validate(
            "
            TYPE A : STRUCT
                b : B;
            END_STRUCT END_TYPE
            
            TYPE B : STRUCT
                c : C;
            END_STRUCT END_TYPE
            
            TYPE C : STRUCT
                c : C;
                e : E;
            END_STRUCT END_TYPE
            
            TYPE E : STRUCT
                c : C;
            END_STRUCT END_TYPE
            ",
        );

        assert_snapshot!(make_readable(&diagnostics));
    }

    #[test]
    fn two_cycles_with_branch() {
        let diagnostics = parse_and_validate(
            "
            TYPE A : STRUCT
                b : B;
            END_STRUCT END_TYPE

            TYPE B : STRUCT
                c : C;
            END_STRUCT END_TYPE

            TYPE C : STRUCT
                e : E;
            END_STRUCT END_TYPE

            TYPE E : STRUCT
                f : F;
            END_STRUCT END_TYPE

            TYPE F : STRUCT
                g : G;
                b : B;
            END_STRUCT END_TYPE

            TYPE G : STRUCT
                h : H;
            END_STRUCT END_TYPE

            TYPE H : STRUCT
                i : I;
            END_STRUCT END_TYPE

            TYPE I : STRUCT
                f : F;
            END_STRUCT END_TYPE
            ",
        );

        assert_snapshot!(make_readable(&diagnostics));
    }
}

mod arrays {
    use insta::assert_snapshot;

    use crate::{test_utils::tests::parse_and_validate, validation::tests::make_readable};

    #[test]
    fn two_cycles_aa_and_aba() {
        let diagnostics = parse_and_validate(
            "
            TYPE A : STRUCT
                a : ARRAY[0..1] OF A;
                b : ARRAY[0..1] OF B;
            END_STRUCT END_TYPE
            
            TYPE B : STRUCT
                a : A;
            END_STRUCT END_TYPE
            ",
        );

        assert_snapshot!(make_readable(&diagnostics));
    }

    #[test]
    fn one_cycle_bcb() {
        let diagnostics = parse_and_validate(
            "
            TYPE A : STRUCT
                b : ARRAY[0..1] OF B;
            END_STRUCT END_TYPE
            
            TYPE B : STRUCT
                c : ARRAY[0..1] OF C;
            END_STRUCT END_TYPE
            
            TYPE C : STRUCT
                b : ARRAY[0..1] OF B;
            END_STRUCT END_TYPE
            ",
        );

        assert_snapshot!(make_readable(&diagnostics));
    }

    #[test]
    fn one_cycle_with_multiple_identical_members_aba() {
        let diagnostics = parse_and_validate(
            "
            TYPE A : STRUCT 
                b1 : ARRAY[0..1] OF B;
                b2 : ARRAY[0..1] OF B;
                b3 : ARRAY[0..1] OF B;
            END_STRUCT END_TYPE

            TYPE B : STRUCT 
                a : A;
            END_STRUCT END_TYPE
            ",
        );

        assert_snapshot!(make_readable(&diagnostics));
    }

    #[test]
    fn one_cycle_aba_output() {
        let diagnostics = parse_and_validate(
            "
            TYPE A : STRUCT
                b : ARRAY [0..1] OF B;
            END_STRUCT END_TYPE

            FUNCTION_BLOCK B
                VAR_OUTPUT
                    a : A;
                END_VAR
            END_FUNCTION_BLOCK
            ",
        );

        assert_snapshot!(make_readable(&diagnostics));
    }

    #[test]
    fn one_cycle_aba_input() {
        let diagnostics = parse_and_validate(
            "
            TYPE A : STRUCT
                b : ARRAY [0..1] OF B;
            END_STRUCT END_TYPE

            FUNCTION_BLOCK B
                VAR_INPUT
                    a : A;
                END_VAR
            END_FUNCTION_BLOCK
            ",
        );

        assert_snapshot!(make_readable(&diagnostics));
    }

    #[test]
    fn two_cycles_with_branch_input() {
        let diagnostics = parse_and_validate(
            "
            FUNCTION_BLOCK A
                VAR_INPUT
                    b : B;
                END_VAR
            END_FUNCTION_BLOCK

            TYPE B : STRUCT
                c : C;
            END_STRUCT END_TYPE

            FUNCTION_BLOCK C
                VAR_INPUT
                    e : E;
                END_VAR
            END_FUNCTION_BLOCK

            TYPE E : STRUCT
                f : ARRAY [0..1] OF F;
            END_STRUCT END_TYPE
            
            FUNCTION_BLOCK F
                VAR_INPUT
                    g : G;
                    b : B;
                END_VAR
            END_FUNCTION_BLOCK
            
            TYPE G : STRUCT
                h : H;
            END_STRUCT END_TYPE
            
            FUNCTION_BLOCK H
                VAR_INPUT
                    i : I;
                END_VAR
            END_FUNCTION_BLOCK

            TYPE I : STRUCT
                f : ARRAY [0..1] OF F;
            END_STRUCT END_TYPE
            ",
        );

        assert_snapshot!(make_readable(&diagnostics));
    }
}

mod functionblocks {
    use insta::assert_snapshot;

    use crate::{test_utils::tests::parse_and_validate, validation::tests::make_readable};

    #[test]
    fn one_cycle_aba_var() {
        let diagnostics = parse_and_validate(
            "
            FUNCTION_BLOCK A
                VAR
                    b : B;
                END_VAR
            END_FUNCTION_BLOCK


            FUNCTION_BLOCK B
                VAR
                    a : A;
                END_VAR
            END_FUNCTION_BLOCK
            ",
        );

        assert_snapshot!(make_readable(&diagnostics));
    }

    #[test]
    fn one_cycle_aba_input() {
        let diagnostics = parse_and_validate(
            "
            FUNCTION_BLOCK A
                VAR_INPUT
                    b : B;
                END_VAR
            END_FUNCTION_BLOCK


            FUNCTION_BLOCK B
                VAR_INPUT
                    a : A;
                END_VAR
            END_FUNCTION_BLOCK
            ",
        );

        assert_snapshot!(make_readable(&diagnostics));
    }

    #[test]
    fn one_cycle_aba_output() {
        let diagnostics = parse_and_validate(
            "
            FUNCTION_BLOCK A
                VAR_OUTPUT
                    b : B;
                END_VAR
            END_FUNCTION_BLOCK


            FUNCTION_BLOCK B
                VAR_OUTPUT
                    a : A;
                END_VAR
            END_FUNCTION_BLOCK
            ",
        );

        assert_snapshot!(make_readable(&diagnostics));
    }

    #[test]
    fn one_cycle_aba_inout() {
        let diagnostics = parse_and_validate(
            "
            FUNCTION_BLOCK A
                VAR_IN_OUT
                    b : B;
                END_VAR
            END_FUNCTION_BLOCK


            FUNCTION_BLOCK B
                VAR_IN_OUT
                    a : A;
                END_VAR
            END_FUNCTION_BLOCK
            ",
        );

        // No recursion because VAR_IN_OUT are treated as pointers
        assert_eq!(diagnostics.len(), 0);
    }

    #[test]
    fn two_cycles_with_branch_input() {
        let diagnostics = parse_and_validate(
            "
            FUNCTION_BLOCK A
                VAR_INPUT
                    b : B;
                END_VAR
            END_FUNCTION_BLOCK

            FUNCTION_BLOCK B
                VAR_INPUT
                    c : C;
                END_VAR
            END_FUNCTION_BLOCK

            FUNCTION_BLOCK C
                VAR_INPUT
                    e : E;
                END_VAR
            END_FUNCTION_BLOCK

            FUNCTION_BLOCK E
                VAR_INPUT
                    f : F;
                END_VAR
            END_FUNCTION_BLOCK
            
            FUNCTION_BLOCK F
                VAR_INPUT
                    g : G;
                    b : B;
                END_VAR
            END_FUNCTION_BLOCK
            
            FUNCTION_BLOCK G
                VAR_INPUT
                    h : H;
                END_VAR
            END_FUNCTION_BLOCK
            
            FUNCTION_BLOCK H
                VAR_INPUT
                    i : I;
                END_VAR
            END_FUNCTION_BLOCK

            FUNCTION_BLOCK I
                VAR_INPUT
                    f : F;
                END_VAR
            END_FUNCTION_BLOCK
            ",
        );

        assert_snapshot!(make_readable(&diagnostics));
    }
}

mod mixed_structs_and_functionblocks {
    use insta::assert_snapshot;

    use crate::{test_utils::tests::parse_and_validate, validation::tests::make_readable};

    #[test]
    fn one_cycle_aba_output() {
        let diagnostics = parse_and_validate(
            "
            TYPE A : STRUCT
                b : B;
            END_STRUCT END_TYPE

            FUNCTION_BLOCK B
                VAR_OUTPUT
                    a : A;
                END_VAR
            END_FUNCTION_BLOCK
            ",
        );

        assert_snapshot!(make_readable(&diagnostics));
    }

    #[test]
    fn one_cycle_aba_input() {
        let diagnostics = parse_and_validate(
            "
            TYPE A : STRUCT
                b : B;
            END_STRUCT END_TYPE

            FUNCTION_BLOCK B
                VAR_INPUT
                    a : A;
                END_VAR
            END_FUNCTION_BLOCK
            ",
        );

        assert_snapshot!(make_readable(&diagnostics));
    }

    #[test]
    fn two_cycles_with_branch_input() {
        let diagnostics = parse_and_validate(
            "
            FUNCTION_BLOCK A
                VAR_INPUT
                    b : B;
                END_VAR
            END_FUNCTION_BLOCK

            TYPE B : STRUCT
                c : C;
            END_STRUCT END_TYPE

            FUNCTION_BLOCK C
                VAR_INPUT
                    e : E;
                END_VAR
            END_FUNCTION_BLOCK

            TYPE E : STRUCT
                f : F;
            END_STRUCT END_TYPE
            
            FUNCTION_BLOCK F
                VAR_INPUT
                    g : G;
                    b : B;
                END_VAR
            END_FUNCTION_BLOCK
            
            TYPE G : STRUCT
                h : H;
            END_STRUCT END_TYPE
            
            FUNCTION_BLOCK H
                VAR_INPUT
                    i : I;
                END_VAR
            END_FUNCTION_BLOCK

            TYPE I : STRUCT
                f : F;
            END_STRUCT END_TYPE
            ",
        );

        assert_snapshot!(make_readable(&diagnostics));
    }
}
