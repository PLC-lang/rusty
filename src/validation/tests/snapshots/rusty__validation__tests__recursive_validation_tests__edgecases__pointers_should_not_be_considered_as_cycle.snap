---
source: src/validation/tests/recursive_validation_tests.rs
expression: diagnostics
---

